src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/CommonInterests.cs

```diff
@@ -0,0 +1,201 @@
+
+            return currentNamespace?.IsGlobalNamespace ?? false;
+        }
+
+#pragma warning disable CA1815 // Override equals and operator equals on value types
+        internal readonly struct QualifiedType
+#pragma warning restore CA1815 // Override equals and operator equals on value types
+        {
+            public QualifiedType(IReadOnlyList<string> containingTypeNamespace, string typeName)
+            {
+                this.Namespace = containingTypeNamespace;
+                this.Name = typeName;
+            }
+
+            public IReadOnlyList<string> Namespace { get; }
+
+            public string Name { get; }
+
+            public bool IsMatch(ISymbol symbol)
+            {
+                return symbol?.Name == this.Name
+                    && symbol.BelongsToNamespace(this.Namespace);
+            }
+
+            public override string ToString() => string.Join(".", this.Namespace.Concat(new[] { this.Name }));
+        }
+
+        internal static class AsyncMethodBuilderAttribute
+        {
+            internal const string TypeName = nameof(System.Runtime.CompilerServices.AsyncMethodBuilderAttribute);
+            internal static readonly IReadOnlyList<string> Namespace = Namespaces.SystemRuntimeCompilerServices;
+        }
+
+        internal static bool HasAsyncCompatibleReturnType([NotNullWhen(true)] this IMethodSymbol? methodSymbol) => IsAsyncCompatibleReturnType(methodSymbol?.ReturnType);
+
+        /// <summary>
+        /// Determines whether a type could be used with the async modifier as a method return type.
+        /// </summary>
+        /// <param name="typeSymbol">The type returned from a method.</param>
+        /// <returns><c>true</c> if the type can be returned from an async method.</returns>
+        /// <remarks>
+        /// This is not the same thing as being an *awaitable* type, which is a much lower bar. Any type can be made awaitable by offering a GetAwaiter method
+        /// that follows the proper pattern. But being an async-compatible type in this sense is a type that can be returned from a method carrying the async keyword modifier,
+        /// in that the type is either the special Task type, or offers an async method builder of its own.
+        /// </remarks>
+        internal static bool IsAsyncCompatibleReturnType([NotNullWhen(true)] this ITypeSymbol? typeSymbol)
+        {
+            if (typeSymbol is null)
+            {
+                return false;
+            }
+
+            // ValueTask and ValueTask<T> have the AsyncMethodBuilderAttribute
+            return (typeSymbol.Name == nameof(Task) && typeSymbol.BelongsToNamespace(Namespaces.SystemThreadingTasks))
+                || IsIAsyncEnumerable(typeSymbol) || typeSymbol.AllInterfaces.Any(IsIAsyncEnumerable)
+                || typeSymbol.GetAttributes().Any(ad => ad.AttributeClass?.Name == AsyncMethodBuilderAttribute.TypeName && ad.AttributeClass.BelongsToNamespace(AsyncMethodBuilderAttribute.Namespace));
```

Comment:
Can't we instead check whether the method is "async", rather than checking whether the return type is attributed?  The compiler already checks all the conditions (which is more than just the attribute) necessary to know whether the return type is valid for use with an async method.  We should be able to then just check whether the method is "async".

Reply:
@stephentoub The purpose of this is to check for methods that have an async-compatible return type, such as `Task` or `ValueTask`, which means it may or may not be an async method (i.e a Task-returning sync method)

Using `typeSymbol.IsAsync` checks whether the method has the `async` keyword, which would produce a different result here.

Reply:
> Using typeSymbol.IsAsync checks whether the method has the async keyword, which would produce a different result here.

Why isn't IsAsync desirable here though? i.e. what would break if we just use IsAsync instead?

Reply:
```csharp
Task Test() {
    Task t = null;
    t.Wait(); // Diagnostic: Use `await t` instead
    return Task.FromResult(1);
}
```
In the above example, using `IsAsync` on method `Test` would return false since it's a sync method. 
The goal of this analyzer is to find sync-blocking invocations (such as `t.Wait()` in this case) while also checking that the parent method or delegate (i.e. `Test`) returns a type that is Async compatible, which has the potential to be made async (such as `Task`, `ValueTask`, `IAsyncEnumberable`, etc...), regardless of whether the method is `async` or not.

Reply:
> using IsAsync on method Test would return false since it's a sync method.

Yes

> which has the potential to be made async

I didn't realize that was a goal for this analyzer as well.  I suspect this is going to hurt more than it helps, in terms of false positives incurred vs real issues found.